test/stablecoin_converter.js

```diff
@@ -0,0 +1,112 @@
+      assert.equal((orderResult.validUntil).toNumber(), 3, "validUntil was stored incorrectly")
+    })
+  })
+  describe("cancelOrder", () => {
+    it("places orders, then cancels it and orders status", async () => {
+      const stablecoinConverter = await StablecoinConverter.new(2**16-1)
+
+      const id = await stablecoinConverter.placeOrder.call(0, 1, true, 3, 10, 20, {from: user_1})
+      await stablecoinConverter.placeOrder(0, 1, true, 3, 10, 20, {from: user_1})
+      const currentStateIndex = await stablecoinConverter.getCurrentStateIndex()
+      await stablecoinConverter.cancelOrder(id, {from: user_1})
+
+      assert.equal(((await stablecoinConverter.orders.call(user_1,id)).validUntil).toNumber(), currentStateIndex - 1, "validUntil was stored incorrectly")
+    })
+  })
+  describe("freeStorageOfOrder", () => {
+    it("places a order, then cancels and deletes it", async () => {
+      const stablecoinConverter = await StablecoinConverter.new(2**16-1)
+
+      const id = await sendTxAndGetReturnValue(stablecoinConverter.placeOrder, 0, 1, true, 3, 10, 20)
+      await stablecoinConverter.cancelOrder(id)
+      await waitForNSeconds(BATCH_TIME)
+      await stablecoinConverter.freeStorageOfOrder(id)
+
+      assert.equal((await stablecoinConverter.orders(user_1, id)).sellAmount, 0, "sellAmount was stored incorrectly")
+    })
+    it("fails to delete non-canceled order", async () => {
+      const stablecoinConverter = await StablecoinConverter.new(2**16-1)
+      const currentStateIndex = await stablecoinConverter.getCurrentStateIndex()
+
+      const id = await sendTxAndGetReturnValue(stablecoinConverter.placeOrder, 0, 1, true, currentStateIndex + 3, 10, 20)
+      await truffleAssert.reverts(stablecoinConverter.freeStorageOfOrder(id), "Order is still valid")
+    })
+    it("fails to delete canceled order in same stateIndex", async () => {
+      const stablecoinConverter = await StablecoinConverter.new(2**16-1)
+    
+      const id = await sendTxAndGetReturnValue(stablecoinConverter.placeOrder, 0, 1, true, 3, 10, 20)
+      await stablecoinConverter.cancelOrder(id)
+      await truffleAssert.reverts(stablecoinConverter.freeStorageOfOrder(id), "Order is still valid")
+    })
+  })
+  describe("addToken()", () => {
+    it("Anyone can add tokens", async () => {
+      const instance = await StablecoinConverter.new(2**16-1)
+
+      const token_1 = await ERC20.new()
+      await instance.addToken(token_1.address)
+
+      assert.equal((await instance.tokenAddressToIdMap.call(token_1.address)).toNumber(), 0)
+      assert.equal(await instance.tokenIdToAddressMap.call(0), token_1.address)
+      const token_2 = await ERC20.new()
+      await instance.addToken(token_2.address)
+
+      assert.equal((await instance.tokenAddressToIdMap.call(token_2.address)).toNumber(), 1)
+      assert.equal(await instance.tokenIdToAddressMap.call(1), token_2.address)
+    })
+
+    it("Reject: add same token twice", async () => {
```

Comment:
This test seems to be more of a test of the `BiMap` contract which already tests this.

Reply:
We shouldn’t assume what the contract is using under the hood. If the top level interface specification says we can’t add a token twice we should have a test for that.